// @remotion/transitions/dist/timings/linear-timing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.linearTiming = void 0;
const remotion_1 = require("remotion");
/**
 * Creates a linear timing object for managing animations in frame units.
 * @description Provides a mechanism to handle frame-based transitions linearly, optionally incorporating easing functions.
 * @see [Documentation](https://remotion.dev/docs/transitions/timings/lineartiming)
 * @param {Object} options Configuration options for the linear timing
 * @param {number} options.durationInFrames Specifies the total duration of the transition in frames
 * @param {((input: number) => number)=} options.easing Optional easing function to modify the interpolation of values
 * @returns {TransitionTiming} An object representing the timing of the transition, including methods to get duration and progress
 */
const linearTiming = (options) => {
    return {
        getDurationInFrames: () => {
            return options.durationInFrames;
        },
        getProgress: ({ frame }) => {
            return (0, remotion_1.interpolate)(frame, [0, options.durationInFrames], [0, 1], {
                easing: options.easing,
                extrapolateLeft: 'clamp',
                extrapolateRight: 'clamp',
            });
        },
    };
};
exports.linearTiming = linearTiming;
